### PR TITLE
Feat: Add the _"wither"_ Pattern to `ServiceBinding`

### PR DESCRIPTION
**Pull request by Johannes-Schneider**
## Context

This PR introduces the _"wither"_ pattern in the `ServiceBinding` interface.
This pattern allows customers to quickly and easily obtain **new** `ServiceBinding` instances with "modified" values.

Example:

```java
ServiceBinding originalBinding;
originalBinding.getServiceIdentifier(); // "my-service"

ServiceBinding changedBinding = originalBinding.withServiceIdentifier(ServiceIdentifier.of("something-else"));
changedBinding.getServiceIdentifier(); // "something-else"

assertThat(originalBinding).isNotSameAs(changedBinding); // true
```

These newly created `ServiceBinding` instances use delegation so that superfluous copying of avioded.

### Feature Scope

- [x] Introduce API 
    - [x] Add default implementation

<details>
<summary><h3>Release Notes</h3></summary>

## Module: `java-core-api`

* Add new API to the `ServiceBinding` interface that allows for easy and convenient access to "changed" `ServiceBinding` instances. These new APIs (`with<X>(...)`) will create a NEW `ServiceBinding` instances that contain exactly the same values as the original instance but with just the provided property overwritten.

</details>

### Definition of Done

- [x] Feature scope is implemented
- [x] Feature scope is tested
- [ ] Feature scope is documented
- [x] Release notes are created

### CODE REVIEW

File: api-parent/core-api/src/main/java/com/sap/cloud/environment/servicebinding/api/DefaultServiceBinding.java [hunk 1]
```diff
@@ -139,7 +139,15 @@ public Map<String, Object> getCredentials()
...
-        return Objects.hash(getName(), getServiceName(), getServicePlan(), getTags(), getCredentials(), properties);
+        return Objects
+            .hash(
+                getName(),
+                getServiceName(),
+                getServiceIdentifier(),
+                getServicePlan(),
+                getTags(),
+                getCredentials(),
+                properties);
```
Comment (Johannes-Schneider) [inline api-parent/core-api/src/main/java/com/sap/cloud/environment/servicebinding/api/DefaultServiceBinding.java:150]: This is a drive-by change since we forgot to include the newly introduced `getServiceIdentifier` into the `equals` and `hashCode` implementation

File: api-parent/core-api/src/main/java/com/sap/cloud/environment/servicebinding/api/ServiceBinding.java [hunk 2]
```diff
@@ -99,4 +100,161 @@ default Optional<ServiceIdentifier> getServiceIdentifier()
...
+     *            The properties to set. Supplying {@code null} is equivalent to supplying an empty {@link Map}.
+     * @return A <b>NEW</b> {@link ServiceBinding} instance that contains the same values as this {@link ServiceBinding}
+     *         but with the given {@code properties}.
+     * @since 0.10.0
+     */
+    @Nonnull
+    default ServiceBinding withProperties( @Nullable final Map<String, Object> properties )
+    {
+        return DelegatingServiceBinding.builder(this).withProperties(properties).build();
+    }
```
Comment (newtork) [inline api-parent/core-api/src/main/java/com/sap/cloud/environment/servicebinding/api/ServiceBinding.java:140]: I feel it's redundant to have both, a wither pattern and a builder pattern 🤔

For an expert library like this, I would expect API minimalism and no usage ambiguity.

Comment (Johannes-Schneider) [reply api-parent/core-api/src/main/java/com/sap/cloud/environment/servicebinding/api/ServiceBinding.java:140]: Please note that the `DelegatingServiceBinding` (including it's builder) is **not** public API.
The customers cannot (as of now) use the `DelegatingServiceBinding.builder(...)` themselves and always have to rely on the `ServiceBinding.with<Foo>(...)` API.

This decision can (and should) be challenge IMO.

As an alternative, for example, we could drop the additional API on the `ServiceBinding` interface (to keep the API as slim as possible) and just make the `DelegatingServiceBinding` class public.
This would come with the downside of more boilerplate code for "manipulating" individual properties:

```java
ServiceBinding originalBinding;
ServiceBinding changedBinding = originalBinding.withName("new name");

vs.

ServiceBinding originalBinding;
ServiceBinding changedBinding = DelegatingServiceBinding.builder(originalBinding).withName("new name").build();
```

Comment (newtork) [reply api-parent/core-api/src/main/java/com/sap/cloud/environment/servicebinding/api/ServiceBinding.java:140]: > the `DelegatingServiceBinding` (including it's builder) is **not** public API.

Could we remove the API then in favor of _wither_ pattern?

Comment (Johannes-Schneider) [reply api-parent/core-api/src/main/java/com/sap/cloud/environment/servicebinding/api/ServiceBinding.java:140]: This library does not contain `Lombok` (to keep the dependency footprint as small as possible).

---

> Could we remove the API then in favor of wither pattern?

Sure. We could remove the `DelegatingServiceBinding.Builder` class, but then we would need to call the constructor, which is easy to mess up due to all the `null`s that need to be passed.

Comment (newtork) [reply api-parent/core-api/src/main/java/com/sap/cloud/environment/servicebinding/api/ServiceBinding.java:140]: Instead of builder pattern, would / have you considered a mutable class with constant delegates and nullable attributes that have a package private setter?

You could still instantiate a new DelegatingServiceBinding for every wither.

Comment (newtork) [reply api-parent/core-api/src/main/java/com/sap/cloud/environment/servicebinding/api/ServiceBinding.java:140]: (We talked about the motivation and actual customer use-case.)

The original issue is that the default implementation of `getServiceIdentifier()` serves us wrong, in scenarios we didn't expect. Our assumption `/service name/ -> /service identifier/` is proven wrong!

And enabling additional API to provide better heuristic(s) to infer `/service identifier/` will be too big of a scope at the moment. Therefore we'll allow for delegating/manipulating copies of `ServiceBinding` instead. This will enable the customer use-case again and buys us time to rethink the logic.

**Comment by Johannes-Schneider** [comment]
Discontinued for the moment.
